Parser: avoid on-demand parsing when parsing interface tokens.

nkcsgexi opened the pull request:
This is to address @rintaro 's comments on the previous PR about on-demand parsing. The interface hash relies on the order of tokens getting consumed; thus on-demand parsing will make the generated hash unstable.

Commit:
Propagate concrete types for all arguments of an apply instruction

lorentey committed:
[stdlib] Implement the new _modify accessor in Dictionary.subscript

This enables in-place mutations of the key-based subscript.
These get rid of one of two full hash table lookup operations (including hashing the key) relative to the getter+setter approach. They can also eliminate COW copies in code like this:

dictionary[foo]?.append(bar)

(Admittedly this isn’t a very useful example.)

lorentey committed:
[stdlib] Fix crucial thinko

gottesmm committed:
[sil-ownership] Tighten up the verification of guaranteed phi arguments.

Now we properly enforce that all guaranteed phi arguments's lifetime is
completely enclosed by the lifetimes of the phi argument's inputs. This is
implemented by inserting an "implicit" regular use into the use stream of the
phi argument's inputs. This is nice since it ensures that when we are verifying
ownership of a specific argument, we will not walk through all
subarguments... instead we just need to verify that the end_borrow is completely
enclosed within the lifetime of the input guaranteed value.

I also fixed a case where we were bailing early out of the SIL Ownership
Verifier causing us to emit an incorrect error. This error would only occur if
we already diagnosed an error.

I also re-added all of the tests that I removed when I replaced
end_borrow_argument with end_borrow.

rdar://33440767

slavapestov committed:
IRGen: Add mangling for method lookup functions

slavapestov committed:
Runtime: Introduce swift_lookUpClassMethod()

slavapestov committed:
IRGen: Emit method lookup function

slavapestov committed:
IRGen: Use method lookup function for resilient super method calls

Fixes <https://bugs.swift.org/browse/SR-3928>, <rdar://problem/31411193>.

compnerd committed:
SwiftDemangle: disable ABI breaking checks

Since this library uses ADT headers, it will pull in a dependency for
LLVMSupport (Error.cpp has the necessary variable declaration for ABI breaking
checks).  Disable the ABI breaking check validation as we only use the library
for header-only dependencies.  Avoiding this dependency shaves nearly 900 KiB
from the resultant binary.

jrose-apple committed:
Add ModuleDecl::ReverseFullNameIterator

Package up the logic that generates a full Clang module name, so that
(a) we don't have to deal with clang::Module in quite as many places
in the /Swift/ compiler, and (b) we can avoid the cost of a temporary
string in a few places.

The main places where this is /not/ adopted is where we don't just
want to know the parent module name, but actually the module itself.
This is mostly indexing-related queries, which use the very similar
ModuleEntity class also defined in Module.h. I didn't quite see an
obvious way to unify these, but that might be where we want to go.

No functionality change.

jrose-apple committed:
When sorting imports for uniquing purposes, use full module names

Fixes a longstanding issue where submodules with the same name in
different top-level modules weren't being sorted deterministically.
This doesn't come up very much in practice, and it would have been
hard to notice anything wrong, but it's good to be right.

slavapestov committed:
IRGen: Don't visit class members when building resilient class metadata layouts

If the class is resilient to us, we cannot assume anything about the
layout of its member area, so let's not even try; just skip visiting
the members entirely.

slavapestov committed:
IRGen: Refuse to compute witness table layouts of resilient protocols

It's not correct to depend on the order of entries in a witness table
if the protocol is resilient, so don't even try.

nkcsgexi committed:
Parser: avoid on-demand parsing when parsing interface tokens.

nkcsgexi commented:
@swift-ci please smoke test

jrose-apple commented:
Can you add a test case? We definitely don't want to regress here!

rintaro commented:
I'm worrying about delayed parsing might randomly updates the interface-hash.
But I think we cannot simply disable it for actual delayed parsing, because if we do that, changes in iterable decl contexts aren't reflected in interface-hash..

rintaro commented on lib/Parse/ParseDecl.cpp:
```diff
@@ -3332,6 +3332,8 @@ bool Parser::parseDeclList(SourceLoc LBLoc, SourceLoc &RBLoc,
 
 bool Parser::canDelayBodyParsing() {
+  if (IsParsingInterfaceTokens)
+    return false;
```
I think this disables "delay parsing" effectively.
`llvm::SaveAndRestore<bool> S(P.IsParsingInterfaceTokens, true)` is at `swift::parseIntoSourceFile()` (it's in `ParseSIL.cpp` but this is the main entry point of parsing)

nkcsgexi replied:
ah, does it mean we always collect interface tokens for hashing and this requires no delayed parsing at all?

rintaro replied:
Yeah, something like this:
```swift
/*<enabled>*/
class C {
  var value = 12 // This affects interface-hash because modifying the initializer might change the type of the property. 

  func foo<T>()  where T : P /*<disabled>*/{
    // Here, no token affects interface-hash.
  }/*</disabled>*/

  func bar() -> Baz /*<disabled>*/{
  }/*</disabled>*/
}
/*</enabled>*/

nkcsgexi replied:
thanks for the example! probably the only parsing can be delayed is for function bodies then :(

nkcsgexi commented:
@jrose-apple is it possible that we generate interface-hash from the AST level rather than the token level? which solves this breakages fundamentally?

airspeedswift committed:
[stdlib] Update Array.subscript to use _modify (#19154)

* Switch Array to use subscript _modify

* Add _modify to ContiguousArray

* XFAIL linux failing test

nkcsgexi commented:
But `P.IsParsingInterfaceTokens` is always set true in ParseSIL.cpp:120. Do we ever set the flag to false in the parser except skipped function bodies? or are you saying we collect these hashes all the time but we don't use them except for primary files?

Commit:
Merge closures_swift4.swift back into closures.swift.

This should have happened when closures_swift3.swift was removed.

jrose-apple commented:
The latter, sorry. We could have parseIntoSourceFile take a flag to decide whether we need it, or put the flag in PersistentParserState (which is already per-SourceFile, right?).

jrose-apple commented:
Or make SourceFile's InterfaceHash Optional, and add a getter to decide if it's wanted or not.

slavapestov committed:
IRGen: Use a method descriptor to identify methods where possible

This eliminates the 'non-resilient class' case here entirely since
all class methods have method descriptors.

slavapestov committed:
IRGen: Remove a few unused methods

nkcsgexi commented:
Can we make parser collect these tokens only when front-end action `-dump-interface-hash` is specified? disregarding of the file under parsing is primary or not?

jrose-apple commented:
No, -dump-interface-hash is a debugging tool. The common use for interface hashes is in reference dependency files.

Commit:
XFAIL array_mutable_assertonly

broken by b4a23adcf060b88af9214220db8c77e7bc611073

rdar://problem/44229005 - for erik

DougGregor committed:
[Loadable-by-address] Replace a never-fails dyn_cast with a cast.

DougGregor committed:
[Associated type inference] Never infer a type involving archetypes and interface types.

Works around an oddity of the associated type inference when the
potential witness is generic and returns (e.g.) Self.

Associated type inference needs to move off of archetypes for the
implementation to become sane. For now, this eliminates a crash.

Fixes rdar://problem/43591024.

nkcsgexi committed:
Parser: calculating interface hash only for primary files.

slavapestov committed:
IRGen: Fix case where we would go through the vtable entry and not the dispatch thunk

Make sure we check resilience of the class defining the method,
and not the class with the override we're calling.

DougGregor committed:
[Type checker] Make sure we have a generic signature when inheriting initializers.

Before adding implicit initializers synthesizes a new initializer, make sure that
we’ve validated the current class declaration to determine it’s generic signature.

Fixes rdar://problem/44235762.

DougGregor committed:
[Associated type inference] Only check for type parameters when bailing out.

Commit:
Store ProtocolDecl rather than TypeLoc for designated protocol.

harlanhaskins committed:
[AST] Remove stored TypeLoc from TypedPattern (#19175)

* [AST] Remove stored TypeLoc from TypedPattern

TypedPattern was only using this TypeLoc as a means to a TypeRepr, which
caused it to store the pattern type twice (through the superclass and through
the TypeLoc itself.)

This also fixes a bug where deserializing a TypedPattern doesn't store
the type correctly and generally cleans up TypedPattern initialization.

Resolves rdar://44144435

* Address review comments

rintaro committed:
[Parse] Disable support for multiline/extended escaping string literal
in attribute message

Strings of diagnostics message processed in EncodedDiagnosticMessage
aren't necessarily from parsed Swift source code. That means, they might
not have quotes around it. Furthermore, memory around them might not be
managed. The logic in 'Lexer::getEncodedStringSegment()' used to cause
access violation.

For now, disable multiline string literal and extended escaping in string
literal for attribute message position. Considering the message might be
from Clang, we cannot simply enable this.

rdar://problem/44228891

DougGregor committed:
[SILGen] Cope with protocol requirement overrides for keypaths.

When emitting accessor calls for keypaths, make sure that we reference
protocol requirements that introduce witness table entries. Other
protocol requirements don't have the necessary dispatch thunks,
resulting in linker errors.

Fixes rdar://problem/44187969.

nkcsgexi committed:
Parser: Ignore and recover marked parse position during on-demand member parsing.

New construction of a parser instance will reset the marked parser
position in PersistentParserState. We should recover the marked parser
position during on-demand parsing in case the marked position will be
later used to create another parser instance to continue parsing
linearly.

DougGregor committed:
[IRGen] Adjust keypath component id for overriding witness methods.

DougGregor committed:
Fix mangled CHECK-SAME line.

rudkx committed:
Revert "Store ProtocolDecl rather than TypeLoc for designated protocol."

Commit:
Disable failing test while investigating.

nkcsgexi committed:
add assertion for no interface hash.
